Feature: Read .TSV Spectral Library In Batches

jcharkow opened the pull request:
## Description

When working with large libraries (such as in-silico libraries), formatting the libraries using `OpenSwathAssayGenerator` and `OpenSwathDecoyGenerator` can be very memory intensive requiring over 100Gb of memory.

This PR is the first which aims to improve memory management through reading the .tsv library file in batches. Note that reading in batches does make the process slower however since library formatting is relatively fast overall I think the tradeoff is worth it. 

Furthermore, upon writing instead of converting everything into a TSV vector and then writing out, each row is pushed individually.  

#### Benchmarks 
Note in this benchmark, I am using a 729Mb library. In-Silico libraries created can be >=10Gb. I am running `TargetedFileConverter` which loads the entire library into memory and then exports it as a TSV. This uses the same methods for loading as `OpenSwathDecoyGenerator` and `OpenSwathAssayGenerator`

Without Using Batches - 4936Mb peak memory usage
![image](https://github.com/user-attachments/assets/210f699f-883f-43c7-929f-0c00cce7872d)

Using Batches (batch size 1 000 000) - 3678Mb peak memory usage
![image](https://github.com/user-attachments/assets/86dedf79-852f-4f37-9427-fbd4796722d1)

Using Batches (batch size 100 000) - 2412Mb peak memory usage
![image](https://github.com/user-attachments/assets/be875018-648b-4aa2-8b1b-b90908533e42)


Note that the entire file is still loaded as an TargetedExperiment into memory since the current algorithms require the entire object to be in memory to work as intended. Since TargetedExperiment inherits from CVParam it can be quite memory hungry so I am also investing reconstructing (or creating an additional class) for TargetedExperiment which is more memory efficient ( see https://github.com/jcharkow/OpenMS/tree/refactorTargetedExp )

Note that this PR is not complete yet (have not added tests) but I wanted to open it as a draft beforehand to get feedback if this is worthwhile before polishing it up further. 


## Checklist
- [ ] Make sure that you are listed in the AUTHORS file
- [ ] Add relevant changes and new features to the CHANGELOG file
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing unit tests pass locally with my changes
- [ ] Updated or added python bindings for changed or new classes (Tick if no updates were necessary.)

### How can I get additional information on failed tests during CI
<details>
  <summary>Click to expand</summary>
If your PR is failing you can check out

- The details of the action statuses at the end of the PR or the "Checks" tab.
- http://cdash.seqan.de/index.php?project=OpenMS and look for your PR. Use the "Show filters" capability on the top right to search for your PR number.
  If you click in the column that lists the failed tests you will get detailed error messages.

</details>

### Advanced commands (admins / reviewer only)
<details>
  <summary>Click to expand</summary>
  
- `/reformat` (experimental) applies the clang-format style changes as additional commit. Note: your branch must have a different name (e.g., yourrepo:feature/XYZ) than the receiving branch (e.g., OpenMS:develop). Otherwise, reformat fails to push.
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds
  
</details>

---
:warning: Note: Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
  - Introduced batch processing options for reading large .tsv files in OpenSwathAssayGenerator, OpenSwathDecoyGenerator, and TargetedFileConverter tools, allowing users to specify a batch size for improved memory efficiency.
  - Added new command-line option "batchSize" to control the number of lines read per batch when importing .tsv files.

- **Bug Fixes**
  - Improved merging of experiments to prevent duplicate entries when combining proteins, peptides, and compounds.

- **Tests**
  - Added new test cases to verify the consistency and correctness of batch processing for .tsv file import and export.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

jcharkow committed:
feature: read TSV file in batches

jcharkow committed:
smarter appending

do not need to check all entries because if sorted than know transitions
will not be duplicated and for peptide groups only the last one would be
possibly duplicated

jcharkow committed:
feature: streamlined output

write feature one at a time instead of storing them in temporary vector
to save memory

jcharkow committed:
use std::move for more memory efficiency

jcharkow committed:
update openswath assay and decoy generator for  batch loading

timosachsenberg commented on src/openms/source/ANALYSIS/OPENSWATH/TransitionTSVFile.cpp:
```diff
@@ -227,8 +227,25 @@ namespace OpenMS
+    }
+    else
+    {
+      endline = offset + batchSize;
+      // stream the offsets to null
+      for (int i = 0; i < offset; ++i)
+      {
+        std::getline(data, line);
+        cnt++;
+      }
```
Do you sweeps through the file multiple times to find the start in line offset? Is this necessary? It might be possible to store the current file pointer position and then use file.seekg(number_of_bytes); to continue reading where one left off - without rereading all lines until offset from disc.

jcharkow commented:
Removing CV term list is probably fine. However this would break traml support.

jcharkow commented:
Current implementation (maybe rename ReactionMonitoringTrasitionTwo) is sufficient for current purposes if needed. Will revisit once apache arrow support is added.

jcharkow committed:
update to use byte offset instead of line offset

this should speed up reading in batches by preventing the need to
constantly reread lines.

jcharkow committed:
add test for reading in batches

jcharkow committed:
Merge branch 'develop' into targetedExpLowMem
